sdk/flow/cli.go

```diff
@@ -3,38 +3,175 @@ package flow
 import (
 	"flag"
 	"os"
+	"reflect"
 	"strings"
+	"sync"
+	"time"
 
 	"github.com/codegangsta/cli"
 	"github.com/codelingo/lingo/app/util"
+	"github.com/golang/protobuf/proto"
 	"github.com/juju/errors"
 )
 
+type DecoratedResult struct {
+
+	// Ctx are the flags and args parsed from the decorated fact
+	Ctx *cli.Context
+
+	// Payload is the struct returned form the flow server for the matched
+	// fact.
+	Payload proto.Message
+}
+
+type flowRunner struct {
+	cliCMD       *CLICommand
+	decoratorCMD *DecoratorCommand
+}
+
+type CLICommand struct {
+	cli.Command
+	Request func(*cli.Context) (chan proto.Message, chan error, func(), error)
+}
+
+type DecoratorCommand struct {
+	cli.Command
+	ConfirmDecorated func(*cli.Context, proto.Message) (bool, error)
+}
+
+func NewFlow(cliCMD *CLICommand, decoratorCMD *DecoratorCommand) *flowRunner {
+
+	return &flowRunner{
+		cliCMD:       cliCMD,
+		decoratorCMD: decoratorCMD,
+	}
+}
+
+func (f *flowRunner) Run() (decoratedResultc chan *DecoratedResult, err error) {
+	defer func() {
+		if err != nil {
+			close(decoratedResultc)
+		}
+	}()
+
+	resultc, errc, cancel, err := f.RunCLI()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// If user is manually confirming reviews, set a long timeout.
+	timeout := time.After(time.Hour * 1)
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		wg.Wait()
+		close(decoratedResultc)
```

Comment:
Won't this panic, because the channel is already closed in a defer at the top of the function?

Reply:
Only if we hit an error. A better pattern is to use a wait chan, I'll update it.